lib/pure/bitops.nim

```diff
@@ -727,89 +730,296 @@ proc countTrailingZeroBits*(x: SomeInteger): int {.inline, noSideEffect.} =
+      0b11111111000000000000000011111111_00000000111111111111111100000000'u64
 
+  when nimvm:
+    rotl(value, shift.int32)
+  else:
+    when useBuiltinsRotate and defined(amd64):
+      when defined(clang):
+        builtin_rotl64(value.culonglong, shift.culonglong).uint64
+      else:
+        builtin_rotl64(value.culonglong, shift.cint).uint64
+    else:
+      rotl(value, shift.int32)
 
-proc rotateRightBits*(value: uint8;
-            amount: range[0..8]): uint8 {.inline, noSideEffect.} =
+func rotateRightBits*(value: uint8, shift: range[0..8]): uint8 {.inline.} =
   ## Right-rotate bits in a 8-bits value.
   runnableExamples:
-    doAssert rotateRightBits(0b0000_0001'u8, 1) == 0b1000_0000'u8
-    doAssert rotateRightBits(0b0000_0001'u8, 2) == 0b0100_0000'u8
-    doAssert rotateRightBits(0b0100_0001'u8, 1) == 0b1010_0000'u8
-    doAssert rotateRightBits(0b0100_0001'u8, 2) == 0b0101_0000'u8
+    doAssert rotateRightBits(0b0110_1001'u8, 1) == 0b1011_0100'u8
+    doAssert rotateRightBits(0b0110_1001'u8, 2) == 0b0101_1010'u8
+    doAssert rotateRightBits(0b0110_1001'u8, 3) == 0b0010_1101'u8
+    doAssert rotateRightBits(0b0110_1001'u8, 4) == 0b1001_0110'u8
 
-  let amount = amount and 7
-  result = (value shr amount) or (value shl ( (-amount) and 7))
+  when nimvm:
+    rotr(value, shift.int32)
+  else:
+    when useBuiltinsRotate:
+      when defined(gcc) or defined(icl):
+        builtin_rotr8(value.cuchar, shift.cint).uint8
+      elif defined(clang):
+        builtin_rotr8(value.cuchar, shift.cuchar).uint8
+      elif defined(vcc):
+        builtin_rotr8(value.cuchar, shift.cuchar).uint8
+    else:
+      rotr(value, shift.int32)
 
-proc rotateRightBits*(value: uint16;
-            amount: range[0..16]): uint16 {.inline, noSideEffect.} =
+func rotateRightBits*(value: uint16, shift: range[0..16]): uint16 {.inline.} =
   ## Right-rotate bits in a 16-bits value.
-  ##
-  ## See also:
-  ## * `rotateRightBits proc <#rotateRightBits,uint8,range[]>`_
-  let amount = amount and 15
-  result = (value shr amount) or (value shl ( (-amount) and 15))
+  runnableExamples:
+    doAssert rotateRightBits(0b00111100_11000011'u16, 1) ==
+      0b10011110_01100001'u16
+    doAssert rotateRightBits(0b00111100_11000011'u16, 3) ==
+      0b01100111_10011000'u16
+    doAssert rotateRightBits(0b00111100_11000011'u16, 6) ==
+      0b00001100_11110011'u16
+    doAssert rotateRightBits(0b00111100_11000011'u16, 8) ==
+      0b11000011_00111100'u16
 
-proc rotateRightBits*(value: uint32;
-            amount: range[0..32]): uint32 {.inline, noSideEffect.} =
+  when nimvm:
+    rotr(value, shift.int32)
+  else:
+    when useBuiltinsRotate:
+      when defined(gcc) or defined(icl):
+        builtin_rotr16(value.cushort, shift.cint).uint16
+      elif defined(clang):
+        builtin_rotr16(value.cushort, shift.cushort).uint16
+      elif defined(vcc):
+        builtin_rotr16(value.cushort, shift.cuchar).uint16
+    else:
```

Comment:
maybe, not sure:

you use this pattern a lot in this file:
```nim
      when defined(gcc) or defined(icl):
        builtin_rotr16(value.cushort, shift.cint).uint16
      elif defined(clang):
        builtin_rotr16(value.cushort, shift.cushort).uint16
      elif defined(vcc):
        builtin_rotr16(value.cushort, shift.cuchar).uint16
```
how about simplify as:
```nim
      builtin_rotr16(value.cushort, ShiftConvRange16(shift)).uint16
```
and then define `ShiftConvRange16` above

Reply:
define it as a template or proc?

Reply:
it can be done with either, it's a CT only thing anyways